File: android/data/domain/src/main/java/com/data/domain/constant/BaseConstant.kt
package com.data.domain.constant

import android.Manifest
import com.data.domain.constant.chat.VisionTypeEnum
import com.data.domain.constant.chat.VisionUploadTypeEnum

class BaseConstant {

    // 常量
    object Constant {
        const val START_DELAY_TIME = 1200L
        // 头像最大大小 200 * 200 = 160 KB
        const val BITMAP_MAX_SIZE_AVATAR: Int = 200
        const val PACKAGE_NAME: String = "com.magicvector"

        const val CHAT_HISTORY_LIMIT_COUNT = 20
        const val MAX_AGENT_NAME_LENGTH = 20
    }

    // http
    object HttpConstant {
        // 请求是否加认证token前缀 最后要在拦截器检查去掉; 定义一些不像url的命名避免出现与后端路由重合
        const val AUTH_TOKEN_PREFIX: String = "/has-0!0-token"

        // 连接超时：2200ms
        const val CONNECT_TIMEOUT = 2200L
        // 读取超时：10s
        const val READ_TIMEOUT = 10_000L
        // 写入超时：10s
        const val WRITE_TIMEOUT = 10_000L
        // 响应处理超时时间：30s
        const val CALL_TIMEOUT = 30_000L
    }

    // url
    object ConstantUrl {
        const val LOCAL_HOST = "192.168.1.2"
        const val TEST_HOST = "192.168.1.2"
        const val LOCAL_ADDRESS = "$LOCAL_HOST:48888"
        const val TEST_ADDRESS = "$TEST_HOST:48888"


        const val LOCAL_URL = "http://$LOCAL_ADDRESS";
        const val TEST_URL = "http://$TEST_ADDRESS";
        const val PROD_URL = "https://api.github.com";

        const val LOCAL_WS_URL = "ws://$LOCAL_ADDRESS";
        const val TEST_WS_URL = "ws://$TEST_ADDRESS";
        const val PROD_WS_URL = "wss://api.github.com";
    }

    object WSConstantUrl {
        const val AGENT_REALTIME_CHAT_URL = "/agent/realtime/chat"
    }

    object PermissionConstant {
        // permission
        val MUST_PERMISSIONS: Array<String> = arrayOf(
            Manifest.permission.CAMERA,
        )
        val NOT_MUST_PERMISSIONS: Array<String> = arrayOf(
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE,
        )
    }

    object NetworkCode {
        const val SUCCESS = 200
        const val SUCCESS_CODE: String = "200"
    }

    object AUDIO {
        const val REALTIME_CHAT_SAMPLE_RATE = 24_000
    }

    object VAD {
        const val SILENCE_DURATION_MS = 100L
    }

    object YOLO {
        const val FILTER_SIZE = 0.1f * 0.1f
        const val PERSON_CLS = 0

        const val OBJECT_SDIFF_W = 0.1f
        const val PERSON_COUNT_W = 0.1f
        const val OBJECT_COUNT_W = 0.05f

        const val PERSON_THRESHOLD_VALUE = 0.10f
        const val OBJECT_THRESHOLD_VALUE = 0.40f
    }

    object VISION {
        // vision上传方法: http, ws-fragment, rtmp
        val UPLOAD_METHOD = VisionUploadTypeEnum.HTTP
        val VISION_TYPE = VisionTypeEnum.IMAGE
        // 分片大小: 兼容ws和mqtt的最大限制
        const val FRAGMENT_SIZE = 16 * 1024 // 16kB大小
        const val WS_SHARD_UPLOAD_DELAY = 20L // 分片上传延迟50ms，避免网络拥塞
    }

    object UDP {
        const val PORT = 45000
        const val CHUNK_SIZE = 4 * 1024
        const val MAX_PACKET_SIZE = 64 * 1024
        const val MIN_FRAME_INTERVAL = 100L
        const val BITMAP_QUALITY = 70
        const val UDP_TIMEOUT = 5000
    }
}

File: app/flutter/lib/domain/constant/base_constant.dart
// 导入枚举
import '../chat/vision_upload_type_enum.dart';
import '../chat/vision_type_enum.dart';

// 常量 - 使用顶层类
class BaseConstant {
  // 私有构造函数，防止实例化
  BaseConstant._();
  // 静态成员，实现树状访问
  static const Constant constant = Constant._();
  static const HttpConstant http = HttpConstant._();
  static const ConstantUrl url = ConstantUrl._();
  static const WSConstantUrl wsUrl = WSConstantUrl._();
  static const NetworkCode networkCode = NetworkCode._();
  static const AudioConstant audio = AudioConstant._();
  static const YoloConstant yolo = YoloConstant._();
  static const VisionConstant vision = VisionConstant._();
  static const UdpConstant udp = UdpConstant._();
}

// 普通常量
class Constant {
  const Constant._(); // 注意这里是 const 构造函数

  static const Constant _instance = Constant._(); // 单例

  static Constant get instance => _instance;

  // 普通常量成员（注意：这些不是 static 的）
  final int startDelayTime = 1_200;
  // 头像最大大小 200 * 200 = 160 KB
  final int bitmapMaxSizeAvatar = 200;
  final int chatHistoryLimitCount = 20;
  final int maxAgentNameLength = 20;
}

// HTTP常量
class HttpConstant {
  const HttpConstant._();

  // 连接超时：2200ms
  final int connectTimeout = 2_200;
  // 读取超时：10s
  final int readTimeout = 10_000;
  // 写入超时：10s
  final int writeTimeout = 10_000;
  // 响应处理超时时间：30s
  final int callTimeout = 30_000;
}

// URL常量
class ConstantUrl {
  const ConstantUrl._();

  final String localHost = "192.168.1.2";
  final String testHost = "192.168.1.2";

  String get localAddress => "$localHost:48888";
  String get testAddress => "$testHost:48888";

  String get localUrl => "http://$localAddress";
  String get testUrl => "http://$testAddress";
  final String prodUrl = "https://api.vector.com";

  String get localWsUrl => "ws://$localAddress";
  String get testWsUrl => "ws://$testAddress";
  final String prodWsUrl = "wss://api.vector.com";
}

// WebSocket URL常量
class WSConstantUrl {
  const WSConstantUrl._();

  final String agentRealtimeChatUrl = "/agent/realtime/chat";
}

// 网络状态码
class NetworkCode {
  const NetworkCode._();

  final int success = 200;
  final String successCode = "200";
}

// 音频常量
class AudioConstant {
  const AudioConstant._();

  final int realtimeChatSampleRate = 24_000;
}

// YOLO常量
class YoloConstant {
  const YoloConstant._();

  final double filterSize = 0.1 * 0.1;
  final int personCls = 0;

  final double objectSdiffW = 0.1;
  final double personCountW = 0.1;
  final double objectCountW = 0.05;

  final double personThresholdValue = 0.10;
  final double objectThresholdValue = 0.40;
}

// 视觉常量
class VisionConstant {
  const VisionConstant._();

  // 这些是变量，可以修改
  final VisionUploadTypeEnum uploadMethod = VisionUploadTypeEnum.http;
  final VisionTypeEnum visionType = VisionTypeEnum.image;
  // 分片大小: 兼容ws和mqtt的最大限制
  final int fragmentSize = 16 * 1024; // 16kB大小
  final int wsShardUploadDelay = 20; // 分片上传延迟20ms，避免网络拥塞
}

// UDP常量
class UdpConstant {
  const UdpConstant._();

  final int port = 45000;
  final int maxPacketSize = 1_450;
  final int minFrameInterval = 100;
  final int bitmapQuality = 70;
  final int udpTimeout = 5_000;
}